server/stream.go

```diff
@@ -1043,15 +1043,26 @@ func (mset *Stream) processInboundJetStreamMsg(_ *subscription, pc *client, subj
 		hdr = msg[:pc.pa.hdr]
 		msg = msg[pc.pa.hdr:]
 	}
-	seq, ts, err = store.StoreMsg(subject, hdr, msg)
-	if err == nil && seq > 0 {
-		mset.lseq = seq
-		mset.lmsgId = msgId
-	}
+
+	// Assume this will succeed.
+	olseq, olmsgId := mset.lseq, mset.lmsgId
+	mset.lseq++
+	mset.lmsgId = msgId
 
 	// We hold the lock to this point to make sure nothing gets between us since we check for pre-conditions.
 	mset.mu.Unlock()
 
+	// Store actual msg.
+	seq, ts, err = store.StoreMsg(subject, hdr, msg)
+
+	// If we did not succedd put those values back.
+	if err != nil || seq == 0 {
+		mset.mu.Lock()
+		mset.lseq = olseq
```

Comment:
Understand that this may completely mess up the state of the stream with concurrent calls to this function and if any of the thread failed to store a message. 

My strong opinion is that I would rather have the original approach with the atomic check (again, although wrong) than this one. As I said, the worse that can happen in the atomic approach is the unprotected map iteration of the consumers... 
Btw, you could always have a dedicated lock to protect the consumer map or use a sync.map - I understand that you use len() so adding/checking in some cases would still require to be done under the mset lock, but for storeUpdates, since it is just iterating, the sync.map itself may be ok).

Reply:
Understood, but was not a fan of the atomic approach. I think we go with this for now.

Reply:
We used to not hold lock for StoreMsg() fyi, I moved where we released the lock just recently due to precondition stuff.

Reply:
Well, there was a reason to move it under lock. I fear that this is going to be really bad if a failure occurs. At least we log the error, so that would be an indication. I see that you do that if `err != nil || seq == 0`. Is there a case where an error is not returned but seq==0 still indicates an error? If so, then in this case we would not even know from the logs..

Reply:
I only moved it under the lock for the preconditions. It has always been outside, which is what I put it back to and assume preconditions will do the right thing. If there are multiple producers the preconditions on a failure could be put back to wrong values (temporarily), but if you have multiple producers you are probably not doing pre-conditions. At least with highly contending multiple producers..

Reply:
So those values are used just in case publishers provide headers that require expected last sequence and/or last message ID? That's the only thing this would affect?

Reply:
Correct.